src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+ep_rt_mono_get_byte_count_in_event(BulkTypeValue *bulk_type_value)
+{
+	uint32_t s_name_len = 0;
+	if (bulk_type_value->s_name)
+		s_name_len = strlen(bulk_type_value->s_name);
+
+	return
+		sizeof(bulk_type_value->fixed_sized_data) +
+		sizeof(bulk_type_value->c_type_parameters) +
+		(s_name_len + 1) * sizeof(char) + 			// Size of name, including null terminator
+		bulk_type_value->c_type_parameters * sizeof(uint64_t);	// Type parameters
+}
+
+// ETW has a limitation of 64K for TOTAL event Size, however there is overhead associated with
+// the event headers.   It is unclear exactly how much that is, but 1K should be sufficiently
+// far away to avoid problems without sacrificing the perf of bulk processing.
+static const uint32_t cb_max_etw_event = 63 * 1024;
+
+// Estimate of how many bytes we can squeeze in the event data for the value struct
+// array.  (Intentionally overestimate the size of the non-array parts to keep it safe.)
+static const uint32_t k_max_bytes_type_values = (cb_max_etw_event - 0x30);
+
+// Estimate of how many type value elements we can put into the struct array, while
+// staying under the ETW event size limit. Note that this is impossible to calculate
+// perfectly, since each element of the struct array has variable size.
+//
+// In addition to the byte-size limit per event, Windows always forces on us a
+// max-number-of-descriptors per event, which in the case of BulkType, will kick in
+// far sooner. There's a max number of 128 descriptors allowed per event. 2 are used
+// for Count + ClrInstanceID. Then 4 per batched value. (Might actually be 3 if there
+// are no type parameters to log, but let's overestimate at 4 per value).
+static const uint32_t k_max_count_type_values = (128 - 2) / 4;
+
+struct _BulkTypeEventLogger {
+	BulkTypeValue m_rg_bulk_type_values[k_max_count_type_values];
+	uint32_t m_n_bulk_type_value_count;
+	uint32_t m_n_bulk_type_value_byte_count;
+	uint8_t *m_p_bulk_type_event_buffer[65536];
+};
+
+//---------------------------------------------------------------------------------------
+//
+// ep_rt_mono_fire_bulk_type_event fires an ETW event for all the types batched so far,
+// it then resets the state to start batching new types at the beginning of the
+// m_rg_bulk_type_values array.
+//
+
+void
+ep_rt_mono_fire_bulk_type_event (BulkTypeEventLogger *p_type_logger)
+{
+	if (p_type_logger->m_n_bulk_type_value_count == 0)
+		return;
+	memset(p_type_logger->m_p_bulk_type_event_buffer, 0, 65536 * sizeof(uint8_t));
+	uint16_t n_clr_instance_id = clr_instance_get_id();
+
+	uint32_t i_size = 0;
+
+	for (int i_type_data = 0; i_type_data < p_type_logger->m_n_bulk_type_value_count; i_type_data++)
+	{
+		BulkTypeValue *target = &p_type_logger->m_rg_bulk_type_values[i_type_data];
+
+		memcpy(p_type_logger->m_p_bulk_type_event_buffer + i_size,
+			   &target->fixed_sized_data,
+			   sizeof(target->fixed_sized_data));
+		i_size += sizeof(target->fixed_sized_data);
+
+		char *wsz_name = target->s_name;
```

Comment:
Right, I naively set it to char as it is [`StackSString`](https://github.com/dotnet/runtime/blob/0e1f46b395052aa917aae377f5cc7f8a5eb07555/src/coreclr/vm/eventtracepriv.h#L237) and [`LPCWSTR`](https://github.com/dotnet/runtime/blob/0e1f46b395052aa917aae377f5cc7f8a5eb07555/src/coreclr/vm/eventtrace.cpp#L1673) in CoreCLR, and at first glance, I didn't know the 100% correct type for that in c.

Reply:
I'm not sure about `StackSString` (I'm guessing it's also wide char), but `LPCWSTR` is shorthand for `Long Pointer Const Wide-char STRing`. The CoreCLR code base is littered with macros like that. In general, `LPC` just means a `const pointer` to whatever the end type is.

Reply:
I'm starting to understand the coreclr types more. [`StackSString`](https://github.com/dotnet/runtime/blob/4cd4cff89cce3d48b70fdb3bfd604d3982d5e808/src/coreclr/inc/sstring.h#L953) seems to be [`InlineSString<512>`](https://github.com/dotnet/runtime/blob/4cd4cff89cce3d48b70fdb3bfd604d3982d5e808/src/coreclr/inc/sstring.h#L807-L808) which seems to function like a byte array of size 512. More implementation details https://github.com/dotnet/runtime/blob/4cd4cff89cce3d48b70fdb3bfd604d3982d5e808/src/coreclr/inc/sstring.inl#L72-L99 and https://github.com/dotnet/runtime/blob/4cd4cff89cce3d48b70fdb3bfd604d3982d5e808/src/coreclr/inc/sbuffer.inl#L20-L45.

Reply:
Yes, it uses a preallocated buffer of 512 bytes, but since it inherits from SString it can also switch to dynamic allocated memory if it runs out of preallocated space. If you want to do something similar you can have a look at  EventPipeWriteEvent* functions that implement something similar, but if you want to reuse it, move it into a struct type, with a preallocated buffer, a pointer to the current buffer and a bool indicating if buffer pointed to is preallocated or dynamic allocated so it can be released, something like:

struct EventPipeStackString {
    ep_char8_t buffer[x];
    ep_char_t * ptr;
    uint32_t size;
    bool free_buffer;
};

and then a couple of functions like:

ep_stack_string_set_capacity (uint32_t size);
ep_stack_string_resize (uint32_t size);
ep_stack_string_get_ptr (uint32_t ensure_size);

and

ep_stack_string_init (EventPipeStackString *instance)
{
    instance->ptr = buffer;
    instance->size = x;
    instance->free_buffer = false;
}

ep_stack_string_fini (EventPipeStackString *instance)
{
    if (instance->free_buffer)
        ep_rt_free_utf8_string (instance->ptr)
    instance->ptr = NULL;
}

You could wrap the definitions above in some macros so you could create some preallocated buffers on some different sizes, 32, 64, 128, 256, 512 for example.

The implementation of ep_stack_string_get_ptr would make sure the buffer is of at least the requested size, switching to dynamic memory if needed.

Usage of the buffer would be to call ep_stack_string_get_ptr with the expected size and then use that pointer for the specific operation.

If you just need the pointer (without need to ensure it has a specific size, like doing strlen, you could implement a getter).

Reply:
Hey folks this is getting really elaborate:

1. Is the string supposed to be UTF8 or UTF16 when it goes into the buffer?  all this talk of `wchar_t` is confusing.  We generally don't use `wchar_t` in Mono - we use `gunichar2` for utf16 and `ep_char8_t*` for UTF8
2. Is it important to use fancy string storage here? Anything wrong with just heap allocating a buffer?
3. If we need to use fast fancy storage, how about using a MonoMempool ?

/cc @lateralusX

Reply:
Above was an example on how StackSString works and how it **could** be implemented on Mono, as an alternative to use regular stack based fixed buffer or heap allocated buffer. If we are not going to use the type logger outside of send method details events, then the pressure on the heap will be less, so for that use case heap allocated buffer might be enough. In CoreCLR the type logger is used for other events as well (like all GC events), so I mainly assume they use StackSString to reduce pressure on heap due to that.

Using a mempool instance during the lifetime of the logger could probably be an alternative as well, since we need to keep the buffers alive until emitting the types bulked by the logger, so all could be freed at the same time.

In EventPipe we use ep_char8_t for UTF8 and ep_char16_t for UTF16.